### ISSUE TITLE
Type variable must be at least 0?

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/cryptol#802]
```
Cryptol> foldl (+) undefined undefined

[error] at <interactive>:1:1--1:6:
  Ambiguous numeric type: type argument 'n' of 'foldl'
  Must be at least: 0
```

This came up in a class where we're using Cryptol. In the class we provide skeleton functions for exercises where some local definitions are elided using Cryptol's `undefined` and the student is expected to fill in the value. I've simplified the issue I was having into the statement above. Cryptol 2.9 is throwing a strange error now. Can type variables _be_ less than 0?

**Comment by weaversa** [comment]
In case it helps, Cryptol 2.8 says
```
  Defaulting type argument 'n' of 'foldl' to 0
```

**Comment by yav** [comment]
I imagine this is just an artifact of the phrasing.  I think it's trying to say that `n` can be anything at all, so we should maybe special case it with a nicer message, as this does look a bit weird.  Perhaps something like "There are no restrictions on `n`"?

Basically it is saying that you must provide the type `n` as it is important (e.g., as addition may overflow so you'd get different results depending on `n`).  Then it tries to be helpful by providing a lower bound on `n`, and since there are no restrictions on it in this case it is saying that the number must be at least 0, which is somewhat confusing.

**Comment by weaversa** [comment]
As a skeleton for teaching, it is nice that it loads into the interpreter without errors in 2.8. The more information we have to put into the exercise, the less interesting it is for the student. I guess that (was) one nice thing about having defaulting.

**Comment by yav** [comment]
Btw my comment above is a bit inaccurate, the `n` is likely the length of the list being folded.

As for the exercises, it sounds like what you need is a way to write *partial* programs, and I wonder if the work on holes might help some with that (I am not sure of the details of it yet though).    In the mean time, some workarounds that come to mind would be to (1) annotate the `undefined` with a type, or (2) make the missing parts into parameters, for example:
```
exercise1 _1 _2 = foldl (+) _1 _2
```

**Comment by robdockins** [comment]
As to the specific error message: yes, we should probably suppress the "must be at least" line if the lower bound is `0`.

**Comment by robdockins** [comment]
I was curious how the state of the `holey` branch (#804)  on this example.  With that branch (which needs quite some updating and a fair amount of additional fit-and-finish work), we currently get:

```
Cryptol> let f = foldl (+) {! !} {! !}
[warning] at <interactive>:1:9--1:14:
  Defaulting type argument 'n' of 'foldl' to 0
Found hole at  <interactive>:1:25--1:30
    Type: [0]?a`848
    Hole is empty
Found hole at  <interactive>:1:19--1:24
    Type: ?a`848
    Hole is empty
Holes:
Cryptol> :t f
f : {a} (Arith a) => a
Holes:
Cryptol> f

Cannot evaluate polymorphic value.
Type: {a} (Arith a) => a
Holes:
Cryptol> f : [8]
{1! !1}
Holes:
  1: <interactive>:1:19--1:24
       Type:  a`848
       Closure:
```

So, unless we made further changes, simply updating this technique would lead to the same error regarding the sequence length.  However, we might be able to consider type variables appearing the types of holes to be not ambiguous and defer selecting them. @david-christiansen, any idea if that could work?

**Comment by david-christiansen** [comment]
In order to support holes, that branch has a reasonable way of evaluating polymorphic things as well as open terms (basically, they get stuck on free variables and start accumulating residualized eliminations). If what you mean by "defer selecting" is to not default type variables that appear in holes, then I think that would probably work just fine (and be preferable to the current behavior). After all, a hole represents a program that is not yet written, and we don't know anything about it yet. It might be worth saying "Not defaulting n to 0 because it occurs in the hole at C.Cry:42:23--42:45" or something to that effect.

As a bit of explanation of the output above, a hole is something that occurs in a source program, identified by its source location (that's the "found hole at" bits, as well as the source location after `1: ` in the bottom part). The value of a hole expression is a *hole instance*, which are numbered. One hole may give rise to multiple instances, because reduction can substitute them into multiple positions. The type of a hole instance is the type of the hole that gave rise to it, and the closure of a hole instance is the value of each free variable at the time that the instance came into existence.

**Comment by robdockins** [comment]
The basic error message issue is fixed via #817. I think we should open a separate issue if we want to continue discussing holes/partial programs/didactic concerns.
